### PR TITLE
pricer: Add pricer to cache latest price

### PR DESCRIPTION
**Pull request by Benzbeeb**
## Implementation details
Add pricer to cache latest price

## Please ensure the following requirements are met before submitting a pull request:
- [x] The pull request is targeted against the correct target branch
- [x] The pull request is linked to an issue with appropriate discussion and an accepted design OR is linked to a spec that describes the work.
- [x] The pull request includes a description of the implementation/work done in detail.
- [ ] The pull request includes any and all appropriate unit/integration tests
- [x] You have added a relevant changelog entry to `CHANGELOG_UNRELEASED.md`
- [x] You have re-reviewed the files affected by the pull request (e.g. using the `Files changed` tab in the Github PR explorer)

### CODE REVIEW

File: chain/pkg/pricecache/pricecache.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+		priceCache: diskv.New(diskv.Options{
+			BasePath:     basePath,
+			Transform:    func(s string) []string { return []string{} },
+			CacheSizeMax: 32 * 1024 * 1024, // 32MB TODO: Make this configurable
+		}),
+	}
+}
+
+func getFilename(symbol string, minCount uint64, askCount uint64) string {
+	return symbol + "," + strconv.FormatUint(minCount, 10) + "," + strconv.FormatUint(askCount, 10)
```
Comment (taobun) [inline chain/pkg/pricecache/pricecache.go]: Using `fmt.Sprintf(...)` looks better

File: chain/cmd/bandd/main.go [hunk 2]
```diff
@@ -23,13 +26,16 @@ import (
...
+	"github.com/bandprotocol/bandchain/chain/saver"
+	"github.com/bandprotocol/bandchain/chain/x/oracle/types"
 )
 
 const (
 	flagInvCheckPeriod        = "inv-check-period"
 	flagWithEmitter           = "with-emitter"
 	flagDisableFeelessReports = "disable-feeless-reports"
 	flagEnableFastSync        = "enable-fast-sync"
+	flagWithSaver             = "with-saver"
```
Comment (taobun) [inline chain/cmd/bandd/main.go]: Pricer

File: chain/pkg/pricecache/pricecache.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+// AddFile saves the given data to a file in HOME/files directory using symbol,minCount,askCount format as filename.
+func (c Cache) AddFile(symbol string, minCount uint64, askCount uint64, multiplier uint64, px uint64, resolveTime int64) error {
```
Comment (taobun) [inline chain/pkg/pricecache/pricecache.go]: SetPrice

File: chain/pkg/pricecache/pricecache.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+		ResolveTime: resolveTime,
+	})
+	if err != nil {
+		return err
+	}
+	return c.priceCache.Write(getFilename(symbol, minCount, askCount), data)
+}
+
+// GetFile loads the file from the file storage. Returns error if the file does not exist.
+func (c Cache) GetFile(filename string) ([]byte, error) {
```
Comment (taobun) [inline chain/pkg/pricecache/pricecache.go]: GetPrice

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/cmd/bandd/main.go [hunk 5]
```diff
@@ -63,6 +69,7 @@ func main() {
 	rootCmd.PersistentFlags().UintVar(&invCheckPeriod, flagInvCheckPeriod, 0, "Assert registered invariants every N blocks")
 	rootCmd.PersistentFlags().String(flagWithEmitter, "", "[Experimental] Use Kafka emitter")
 	rootCmd.PersistentFlags().Bool(flagEnableFastSync, false, "[Experimental] Enable fast sync mode")
+	rootCmd.PersistentFlags().String(flagWithPricer, "", "[Experimental] Enable mode to save price in price cahce")
```
Comment (taobun) [inline chain/cmd/bandd/main.go]: cache

File: chain/pkg/pricecache/pricecache.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
+package pricecache
+
+import (
+	"fmt"
+
+	"github.com/bandprotocol/bandchain/chain/pkg/obi"
```
Comment (taobun) [inline chain/pkg/pricecache/pricecache.go]: order

File: chain/pkg/pricecache/pricecache.go [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+	}
+}
+
+func GetFilename(symbol string, minCount uint64, askCount uint64) string {
+	return fmt.Sprintf("%s,%v,%v", symbol, minCount, askCount)
+}
+
+// SetPrice saves the given data to a file in HOME/files directory using symbol,minCount,askCount format as filename.
+func (c Cache) SetPrice(symbol string, minCount uint64, askCount uint64, price Price) error {
+	data, err := obi.Encode(price)
```
Comment (taobun) [inline chain/pkg/pricecache/pricecache.go]: Can use MustEncode?

File: chain/x/oracle/client/rest/query.go [hunk 8]
```diff
@@ -128,6 +135,41 @@ func getRequestSearchHandler(cliCtx context.CLIContext, route string) http.Handl
...
+		decoder := json.NewDecoder(r.Body)
+		var requestPrices RequestPrices
+		err := decoder.Decode(&requestPrices)
+		if err != nil {
+			rest.WriteErrorResponse(w, http.StatusBadRequest, err.Error())
+			return
+		}
+		prices := make([]pricecache.Price, len(requestPrices.Symbols))
+		height := int64(0)
+		for idx, symbol := range requestPrices.Symbols {
```
Comment (taobun) [inline chain/x/oracle/client/rest/query.go:150]: If we request 25 or 50 symbols in one request how long it takes?

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/x/oracle/client/rest/query.go (hunk 8 above)
Comment (Benzbeeb) [reply chain/x/oracle/client/rest/query.go:150]: 50 symbols = 40ms

File: chain/pkg/pricecache/pricecache.go [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+			CacheSizeMax: 32 * 1024 * 1024, // 32MB TODO: Make this configurable
+		}),
+	}
+}
+
+func GetFilename(symbol string, minCount uint64, askCount uint64) string {
+	return fmt.Sprintf("%s,%v,%v", symbol, minCount, askCount)
+}
+
+// SetPrice saves the given data to a file in HOME/files directory using symbol,minCount,askCount format as filename.
```
Comment (taobun) [inline chain/pkg/pricecache/pricecache.go]: Fix comment

File: chain/pricer/app.go [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+	abci "github.com/tendermint/tendermint/abci/types"
+	"github.com/tendermint/tendermint/libs/log"
+	dbm "github.com/tendermint/tm-db"
+
+	bandapp "github.com/bandprotocol/bandchain/chain/app"
+	"github.com/bandprotocol/bandchain/chain/pkg/pricecache"
+	"github.com/bandprotocol/bandchain/chain/x/oracle/types"
+)
+
+// App extends the standard Band Cosmos-SDK application with Kafka emitter
```
Comment (taobun) [inline chain/pricer/app.go]: Fix comment

File: chain/pricer/app.go [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+		stdOs[oid] = true
+	}
+	return &App{
+		BandApp: app,
+		StdOs:   stdOs,
+		cahce:   pricecache.New(fileDir),
+	}
+}
+
+// EndBlock calls into the underlying EndBlock and emits relevant events to Kafka.
```
Comment (taobun) [inline chain/pricer/app.go]: ditto

File: chain/pricer/app.go [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+	"github.com/bandprotocol/bandchain/chain/pkg/pricecache"
+	"github.com/bandprotocol/bandchain/chain/x/oracle/types"
+)
+
+// App extends the standard Band Cosmos-SDK application with Kafka emitter
+// functionality to act as an event producer for all events in the blockchains.
+type App struct {
+	*bandapp.BandApp
+	StdOs map[types.OracleScriptID]bool
+	cahce pricecache.Cache
```
Comment (taobun) [inline chain/pricer/app.go]: cache

File: chain/pricer/utils.go [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+type EvMap map[string][]string
+
+// JsDict is a type alias for JSON dictionary.
+type JsDict map[string]interface{}
+
+// Message is a simple wrapper data type for each message published to Kafka.
+type Message struct {
+	Key   string
+	Value JsDict
+}
```
Comment (taobun) [inline chain/pricer/utils.go]: Remove unused data type

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/pricer/app.go [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+	return abci.ResponseQuery{
+		Code:      code,
+		Codespace: space,
+		Log:       log,
+		Height:    height,
+		Value:     value,
+	}
+}
+
+func (app *App) Query(req abci.RequestQuery) abci.ResponseQuery {
```
Comment (taobun) [inline chain/pricer/app.go:76]: Add comment

File: chain/pricer/app.go [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+// EndBlock calls into the underlying EndBlock and save price to cache
+func (app *App) EndBlock(req abci.RequestEndBlock) abci.ResponseEndBlock {
+	res := app.BandApp.EndBlock(req)
+	for _, event := range res.Events {
+		app.handleBeginBlockEndBlockEvent(event)
+	}
+	return res
+}
+
+func QueryResultError(err error) abci.ResponseQuery {
```
Comment (taobun) [inline chain/pricer/app.go]: Make it internal function if we only use in this file.

**Review by taobun — CHANGES_REQUESTED** [review]

**Review by taobun — APPROVED** [review]

**Pull request merged**